compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
       TR::Node * skipedConversion = NULL;
       TR::Compilation *comp = cg->comp();
 
-      // Transform land into RISBG as it may be better than discrete pair of ands
+      // Given the right case, transform land into RISBG as it may be better than discrete pair of ands
       if (node->getOpCodeValue() == TR::land && secondChild->getOpCode().isLoadConst())
          {
-         uint64_t value = secondChild->getLongInt();
+         uint64_t longConstValue = secondChild->getLongInt();
 
          if ((firstChild->getOpCodeValue() == TR::i2l || firstChild->getOpCodeValue() == TR::iu2l)
                && firstChild->getReferenceCount()==1 && firstChild->getRegister() == NULL && firstChild->isHighWordZero() )
```

Comment:
`firstChild->getReferenceCount()==1 && firstChild->getRegister() == NULL` can be replaced by `firstChild->isSingleRefUnevaluated()`. Why does the reference count matter here?

Reply:
So the way I understand is that we would originally skip the evaluation of the `i2l` or `iu2l` node if this entire condition holds true. `firstChild` would be set to firstChild->getFirstChild() and if a RISBG instruction can be generated, then we wouldn't evaluate the `i2l` or `iu2l` node anymore. If the reference count is greater than 1 then we cannot skip the evaluation because we might need the result later on. 

I left this check here because my initial thought was that a case like this should never reach the code generator because the optimizer should have further simplified a tree like this. Is this not true? (I can see why the name of the counter can be misleading. I'll try coming up with a better one).

Reply:
(_Outsider Opinion_ If it's expected that the optimizer **must** have done a simplification, we need to documented in some fashion to avoid tripping people up when doing investigations, or when disabling some or all of the optimizer. Right now, there's not a great home for this, but we're working on developing a 'pre-codegen validation' which is where you could sock these kinds of facts away. See #1482 )

Reply:
> So the way I understand is that we would originally skip the evaluation of the i2l or iu2l node if this entire condition holds true.

The function `tryToReplaceLongAndWithRotateInstruction` as it stands does not skip the evaluation of any children. If a truncation operation should not happen a simplifier case should be added to eliminate it. The function implemented here is responsible for determining if an evaluation of an AND node can benefit from being evaluated with an `RISBG` instruction. The reference count of the children of the AND operation should not matter here. Is the logic here correct?